### PR TITLE
KAFKA-16587: Add subscription model information to group state

### PR DESCRIPTION
**Pull request by rreddy-22**
In the Uniform Assignor, we invoke either of the two assignment builders depending on the subscription type. The two types of subscriptions are:
**Homogeneous Subscription Type**

- All the members of the consumer group are subscribed to the same set of topics.
- In this case the Optimized Uniform Assignment Builder is invoked

**Heterogeneous Subscription Type**

- Not all the members of the consumer group are subscribed to the same set of topics.
- In this case the General Uniform Assignment Builder is invoked

During performance testing we realized that checking for the current subscription model on every rebalance is the most time consuming task.
When all the subscriptions are equal, we iterate through every member and check equality for the topic lists. This is very expensive especially when there’s more members.

To combat this we want to store this information in the Consumer Group state and pass it to the assignor via the Target Assignment Builder.

### CODE REVIEW

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: Should we add unit test to validate this logic?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/ConsumerGroupSubscriptionModel.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/ConsumerGroupSubscriptionModel.java]: nit: Should we call it `SubscriptionType`?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/ConsumerGroupSubscriptionModel.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/ConsumerGroupSubscriptionModel.java]: nit: Could we add some javadoc to both of them? I would actually move the description about the two types from the top level javadoc to here.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/ConsumerGroupSubscriptionModel.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/ConsumerGroupSubscriptionModel.java]: nit: Let's add an empty line before this one.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java]: I wonder if we could change the return type of the method from `Map<Uuid, List<String>>` to `Map<Uuid, Collection<String>>` and avoid this copy here. It seems possible because we only iterate over the member ids later on. This could be a nice performance improvement too while we are here.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: nit: I think that we can drop the `group` prefix here. Based on my previous commit, we could also use `subscriptionType` here.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: ditto about the naming.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: Hum... We need to be careful here because  we are not suppose to update the internal state of the group on this code path. It may be better to keep it as it was before and to have another method to update the subscription type.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: nit: We don't need this variable in my opinion. How about the following?

```
int numOfMembers = members.size();
for (Map.Entry<String, Integer> entry : subscribedTopicNames.entrySet()) {
    if (entry.getValue() != numOfMembers) {
        groupSubscriptionModel.set(HETEROGENEOUS);
        return;
    }
}
groupSubscriptionModel.set(HOMOGENEOUS);
```

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilderTest.java
Comment (dajac) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilderTest.java]: nit: As we use the same value everywhere, I would directly inline `HETEROGENEOUS`.

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilderTest.java
Comment (dajac) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/OptimizedUniformAssignmentBuilderTest.java]: ditto.

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/RangeAssignorTest.java
Comment (dajac) [inline group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/RangeAssignorTest.java:692]: Is this change enough to cover the `HETEROGENEOUS` path?

**Review by dajac — COMMENTED** [review]
@rreddy-22 Thanks for the PR. I left a few minor comments. Could you please run the relevant benchmarks before/after this change?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/RangeAssignor.java]: Thanks! Made the change

File: group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/RangeAssignorTest.java
Comment (rreddy-22) [reply group-coordinator/src/test/java/org/apache/kafka/coordinator/group/assignor/RangeAssignorTest.java:692]: Yeah it covers it. Nothing changed in the hetero path anyways

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: This seams incorrect to me. This method is used in consumerGroupHeartbeat method so it should not alter the state of the group. The state is altered when records are replayed. I suppose that we can simply remove it from here.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: nit: I would remove the `maybe` prefix from the name because we effectively always update it.

File: jmh-benchmarks/src/main/java/org/apache/kafka/jmh/assignor/ServerSideAssignorBenchmark.java
Comment (dajac) [inline jmh-benchmarks/src/main/java/org/apache/kafka/jmh/assignor/ServerSideAssignorBenchmark.java:201]: nit: It looks like that I missed this one.

**Review by dajac — COMMENTED** [review]
@rreddy-22 Thanks for the update. I left two more comments. I wait on the results of the benchmarks before merging it.

File: jmh-benchmarks/src/main/java/org/apache/kafka/jmh/assignor/ServerSideAssignorBenchmark.java
Comment (rreddy-22) [reply jmh-benchmarks/src/main/java/org/apache/kafka/jmh/assignor/ServerSideAssignorBenchmark.java:201]: hmm this is weird lol, I could've sworn I removed em all

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: ohhhh that makes sense, I wasn't sure what the reasoning was, okie I'll remove it, thanks for explaining

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: hmm theoretically yes, but what if it stays homo/hetero even after the update? maybe it doesn't matter cause the variable is reassigned

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: This is incorrect, I think. We should initialize it with the current subscribed topic names. Otherwise, we will use an empty map later one. I wonder if we should also call it `subscribedTopicNames` to be consistent.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: nit: Should we call it `computeSubscribedTopicNames` to follow `computeSubscriptionMetadata`?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:1377]: nit: Let's keep these comments at the top of the if branch.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: I would prefer to have this right after calling `updateSubscribedTopicNames` in order to keep all the updates together. I would also call it `computeSubscriptionType`.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: Note that the number of members may be incorrect here. For instance, when a new member joins, it is not accounted into it yet. I wonder if we could do the same without it. Or, we need to adjust it base on whether the member is new or not.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: Could we reuse `maybeUpdateSubscribedTopicNames`?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: I don't get this one. If the subscriptions are empty, isn't the group HOMOGENEOUS by definition?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: nit: Could we keep it private?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: An alternative may be to compare the counts within the Map without considering the group size. This could work because we do not accept empty subscriptions from a member.

**Review by dajac — COMMENTED** [review]
@rreddy-22 Thanks for the update. I left more comments.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: I initialized it in the method to update, and we return the map. Currently there is no getter for the map, I can create another getter to return the map and not just the list of subscribed topic names

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: I figured we don't need to compute it if it's not necessary since we aren't updating the actual value anyways. It is only required in this if block when the target assignment builder is initialized right right

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: I think it was originally named like that since we pass a list of subscribedTopic names in the heartbeat

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: yeah I was actually thinking the same, I saw that the member map is updated in the getOrCreateMember step so I figured the memberCount is updated, unless a member is removed, I'm not sure if that updates elsewhere, or I might be overlooking something

Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: An alternative may be to compare the counts within the Map without considering the group size. This could work because we do not accept empty subscriptions from a member. -> this works as well

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: discussed offline

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:1371]: nit: Should we use `subscribedTopicNames` too?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: nit: Could we move it to right before `subscribedTopicNamesMap = group.computeSubscribedTopicNames(member, updatedMember);`?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: nit: We could remove this as it is already in the `@return`.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java:1008]: nit: Let's add an empty line before this one in order to match the style in the file.

**Review by dajac — COMMENTED** [review]
@rreddy-22 Thanks for the update. I left a few nits. We should be good to go when they are addressed.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:1371]: we can't cause the list is also named subscribedTopicNames right?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: I thought you wanted it above the if statement in the last comment haha, I think it was there originally

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (dongnuo123) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: Do we need to check if referenceCount is equal to member count?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: Sorry. I was not clear in my previous comment.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: We don't allow subscribing with no topic names with the consumer protocol so this logic works in this case. However, thinking about https://github.com/apache/kafka/pull/15798, it may not be true when the classic protocol is used to join a consumer group. @dongnuo123 Do you confirm?

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: I was thinking a little more about this one and I still believe that we should rather do it right after computing `subscribedTopicNamesMap = group.computeSubscribedTopicNames(member, updatedMember);` because there are cases where we would end up here but without needing to re-compute it (e.g. staticMemberReplaced is true).

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java:1371]: While here, could we move those two [lines](https://github.com/apache/kafka/pull/15785/files#diff-00f0f81cf13e66781777d94f7d2e68a581663385c37e98792507f2294c91bb09R1275) to here. It makes sense to keep all those together.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java
Comment (dongnuo123) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: > thinking about https://github.com/apache/kafka/pull/15798, it may not be true

Yes this isn't true. I can add a check to prevent empty subscribed topic list form joining.

Comment (dongnuo123) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: For checking the referenceCount, is it possible to have something like
```
member A - topic-1, topic-2
member B - topic-2, topic-3
member C - topic-1, topic-3
```

Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: ah yes! I remember now that this is why I chose to do total number of members in the group

Comment (rreddy-22) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: we need to somehow get the accurate total member count and use this only

Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: > > thinking about #15798, it may not be true
> 
> Yes this isn't true. I can add a check to prevent empty subscribed topic list form joining.

This is not possible as we must support the current protocol as it is.

Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: > For checking the referenceCount, is it possible to have something like
> 
> ```
> member A - topic-1, topic-2
> member B - topic-2, topic-3
> member C - topic-1, topic-3
> ```

Good point. So we need to use the group size.

Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java]: > we need to somehow get the accurate total member count and use this only

We can get the current count and update it locally when a new member is created and we don't replace a static member, I suppose.

File: group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java
Comment (dajac) [inline group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: Once https://github.com/apache/kafka/pull/15847 is merged, I think that we can easily compute the numMembers as follow:

```
int numMembers = group.numMembers();
if (!group.hasMember(updatedMember.memberId() && !staticMemberReplaced) {
   numMembers++;
}
```

If the member did not exist and we are not replacing a static member, we increment by one.

Comment (dajac) [reply group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java]: I just merged it.

**Review by dajac — APPROVED** [review]
LGTM, thanks.

**Pull request merged**
